### PR TITLE
ARROW-369: [Python] Convert multiple record batches at once to Pandas

### PR DESCRIPTION
**Pull request by BryanCutler**
Modified Pandas adapter to handle columns with multiple chunks with `ConvertColumnToPandas`.  This modifies the pyarrow public API by adding a class `RecordBatchList` and static method `toPandas` which takes a list of Arrow RecordBatches and outputs a Pandas DataFrame. 

Adds unit test in test_table.py to do the conversion for each column with typed specialization.

### CODE REVIEW

**Comment by BryanCutler** [comment]
ping @wesm.  I still need to check that the RecordBatch schemas are equal and add some negative unit tests, but I wanted to check with you if the approach so far looks good.  Could you please take a look when you get a chance?  Thanks!

File: python/pyarrow/table.pyx [hunk 1]
```diff
@@ -296,6 +296,56 @@ cdef class RecordBatch:
...
+                c_array_vec[i].push_back(arr.sp_array)
+
+        # create columns from the chunks
+        names = []
+        data = []
+        for i in range(K):
+            c_col.reset(new CColumn(schema.sp_schema.get().field(i), c_array_vec[i]))
+            # TODO - why need PyOject ref? arr is placeholder
+            check_status(pyarrow.ConvertColumnToPandas(
+                c_col, <PyObject*> arr, &np_arr))
```
Comment (BryanCutler) [inline python/pyarrow/table.pyx]: the `PyObject *` param doesn't seem to do anything, is it needed?

File: python/src/pyarrow/adapters/pandas.cc [hunk 2]
```diff
@@ -684,27 +703,45 @@ class ArrowDeserializer {
...
-        arr.get());
-
-    const T* in_values = reinterpret_cast<const T*>(prim_arr->data()->data());
+    if (data->num_chunks() == 1 && data->null_count() == 0) {
+      return ConvertValuesZeroCopy<TYPE>(data->chunk(0));
+    }
 
-    if (arr->null_count() > 0) {
+    if (data->null_count() > 0) {
       RETURN_NOT_OK(AllocateOutput(NPY_FLOAT64));
```
Comment (BryanCutler) [inline python/src/pyarrow/adapters/pandas.cc]: Just to make sure, an Integer column only needs to be casted to a double if there are Null values right?

File: python/pyarrow/table.pyx [hunk 3]
```diff
@@ -296,6 +296,56 @@ cdef class RecordBatch:
         return result
 
 
+cdef class RecordBatchList:
+    '''
+    A utility class used for conversions with RecordBatch lists
+    '''
+
+    @staticmethod
+    def to_pandas(batches):
```
Comment (wesm) [inline python/pyarrow/table.pyx]: This is a Java-like construct -- just make this a plain old function "dataframe_from_batches" or something like that

File: python/pyarrow/table.pyx (hunk 1 above)
Comment (wesm) [inline python/pyarrow/table.pyx]: I think you can remove the cast. 

If you instead construct a Table from a list of RecordBatches you can avoid the code duplication

File: python/pyarrow/table.pyx [hunk 4]
```diff
@@ -296,6 +296,56 @@ cdef class RecordBatch:
...
+
+        # make a vector of ArrayVectors to store column chunks
+        for i in range(K):
+            c_array_vec.push_back(vector[shared_ptr[CArray]]())
+
+        # copy each batch into a chunk
+        for batch in batches:
+            for i in range(K):
+                arr = batch[i]
+                c_array_vec[i].push_back(arr.sp_array)
```
Comment (wesm) [inline python/pyarrow/table.pyx]: If you instead loop over column index, then loop over batches, you can avoid this `c_array_vec` business and construct the Column directly. See Table.to_pandas

File: python/src/pyarrow/adapters/pandas.cc (hunk 2 above)
Comment (wesm) [inline python/src/pyarrow/adapters/pandas.cc]: right

File: python/src/pyarrow/adapters/pandas.cc [hunk 5]
```diff
@@ -654,27 +650,50 @@ class ArrowDeserializer {
...
+    }
+
+    RETURN_NOT_OK(AllocateOutput(arrow_traits<T2>::npy_type));
+
+    for (int c = 0; c < data->num_chunks(); c++) {
+      const std::shared_ptr<Array> arr = data->chunk(c);
+      arrow::PrimitiveArray* prim_arr = static_cast<arrow::PrimitiveArray*>(
+          arr.get());
+      const T* in_values = reinterpret_cast<const T*>(prim_arr->data()->data());
+      T* out_values = reinterpret_cast<T*>(PyArray_DATA(out_)) + chunk_offset;
```
Comment (wesm) [inline python/src/pyarrow/adapters/pandas.cc]: you're welcome to use `auto` with these casts for better DRY

**Review by wesm — CHANGES_REQUESTED** [review]
This seems fine so far, minor comments

File: python/pyarrow/tests/test_table.py [hunk 6]
```diff
@@ -50,6 +50,41 @@ def test_recordbatch_from_to_pandas():
...
+def test_recordbatchlist_errors():
+    data1 = pd.DataFrame({'c1': np.array([1], dtype='uint32')})
+    data2 = pd.DataFrame({'c1': np.array([4.0, 5.0], dtype='float64')})
+
+    batch1 = pa.RecordBatch.from_pandas(data1)
+    batch2 = pa.RecordBatch.from_pandas(data2)
+
+    # TODO: Enable when subclass of unittest.testcase
+    #with self.assertRaises(pyarrow.ArrowException):
+    #    self.assertRaises(pa.dataframe_from_batches([batch1, batch2]))
```
Comment (BryanCutler) [inline python/pyarrow/tests/test_table.py]: Can these tests be changed to a unittest.testcase subclass?  I could open another JIRA to do this if it's ok

**Comment by BryanCutler** [comment]
Thanks for the review @wesm , I fixed the issues you pointed out and added a check that the schema in each batch is equal.

File: python/pyarrow/table.pyx [hunk 7]
```diff
@@ -296,6 +297,53 @@ cdef class RecordBatch:
...
+        Schema schema_comp
+
+    import pandas as pd
+
+    schema = batches[0].schema
+
+    # check schemas are equal
+    for i in range(1, len(batches)):
+        schema_comp = batches[i].schema
+        if not schema.sp_schema.get().Equals(schema_comp.sp_schema):
```
Comment (BryanCutler) [inline python/pyarrow/table.pyx]: Should the method `CSchema.Equals` also be in the pyarrow class?

Comment (wesm) [inline python/pyarrow/table.pyx]: definitely, pyarrow.schema.Schema.equals

File: python/pyarrow/tests/test_table.py (hunk 6 above)
Comment (wesm) [inline python/pyarrow/tests/test_table.py]: I don't have a strong opinion -- I think we've been trying to use pytest (i.e. fixtures over TestCase classes), @xhochy do you have an opinion?

Comment (BryanCutler) [inline python/pyarrow/tests/test_table.py]: pytest is fine too.  I saw unittest being used somewhere, so I wasn't sure if there was a preference.

**Comment by BryanCutler** [comment]
@wesm, hopefully this is good to go now.  Please take another look when you can, thanks!

**Review by xhochy — APPROVED** [review]
+1, LGTM

File: python/pyarrow/table.pyx [hunk 8]
```diff
@@ -414,6 +415,52 @@ cdef class RecordBatch:
...
+        Schema schema
+
+    import pandas as pd
+
+    schema = batches[0].schema
+
+    # check schemas are equal
+    if any((not schema.equals(other.schema) for other in batches[1:])):
+        raise ArrowException("Error converting list of RecordBatches to "
+                "DataFrame, not all schemas are equal")
```
Comment (wesm) [inline python/pyarrow/table.pyx]: Later we'll want to display the mismatched schemas in the error message but this is ok for now

**Review by wesm — COMMENTED** [review]
+1

File: python/pyarrow/table.pyx (hunk 8 above)
Comment (BryanCutler) [inline python/pyarrow/table.pyx]: Yeah, I should have added that.. I'll make a note to do that later.

**Comment by BryanCutler** [comment]
Thanks for the reviews @xhochy and @wesm!

**Comment by wesm** [comment]
I was thinking a better API for this would be:

```
table = pa.Table.from_batches(batches)
df = table.to_pandas()
```

what do you think?

**Comment by BryanCutler** [comment]
yeah, it would be more flexible that way.  I can go ahead and make a PR for it.

**Pull request closed without merging**